pkg/provider/pulumi/pulumi.go

```diff
@@ -16,49 +16,115 @@ package pulumi
-	errUnmarshalSecret               = "unable to unmarshal secret: %w"
 	errUnableToGetValues             = "unable to get value for key %s: %w"
 	errGettingAllSecretsNotSupported = "getting all secrets is currently not supported by Pulumi"
+	errReadEnvironment               = "error reading environment : %w"
+	errPushSecrets                   = "error pushing secret: %w"
+	errInterfaceType                 = "interface{} is not of type map[string]interface{}"
 )
 
 var _ esv1beta1.SecretsClient = &client{}
 
-func (c *client) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
-	x, _, err := c.escClient.OpenEnvironment(ctx, c.organization, c.environment, "", 5*time.Minute)
+func (c *client) GetSecret(_ context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	env, err := c.escClient.OpenEnvironment(c.authCtx, c.organization, c.environment)
 	if err != nil {
 		return nil, err
 	}
-	value, err := c.escClient.GetOpenProperty(ctx, c.organization, c.environment, x, ref.Key)
+	value, _, err := c.escClient.ReadEnvironmentProperty(c.authCtx, c.organization, c.environment, env.GetId(), ref.Key)
 	if err != nil {
 		return nil, err
 	}
-	return utils.GetByteValue(value.ToJSON(false))
+	return utils.GetByteValue(value.GetValue())
 }
 
-func (c *client) PushSecret(_ context.Context, _ *corev1.Secret, _ esv1beta1.PushSecretData) error {
-	return errors.New(errPushSecretsNotSupported)
+func (c *client) PushSecret(_ context.Context, secret *corev1.Secret, data esv1beta1.PushSecretData) error {
+	value := secret.Data[data.GetSecretKey()]
+
+	updatePayload := &esc.EnvironmentDefinition{
+		Values: &esc.EnvironmentDefinitionValues{
+			AdditionalProperties: map[string]interface{}{
+				data.GetRemoteKey(): string(value),
+			},
+		},
+	}
+	_, oldValues, err := c.escClient.OpenAndReadEnvironment(c.authCtx, c.organization, c.environment)
+	if err != nil {
+		return fmt.Errorf(errReadEnvironment, err)
+	}
+	updatePayload.Values.AdditionalProperties = mergeMaps(oldValues, updatePayload.Values.AdditionalProperties)
+	_, err = c.escClient.UpdateEnvironment(c.authCtx, c.organization, c.environment, updatePayload)
+	if err != nil {
+		return fmt.Errorf(errPushSecrets, err)
+	}
+
+	return nil
+}
+
+func mergeMaps(map1, map2 map[string]interface{}) map[string]interface{} {
+	mergedMap := make(map[string]interface{})
+
+	// Helper function to merge nested maps
+	var mergeNestedMap func(m map[string]interface{}, keys []string, value interface{})
+	mergeNestedMap = func(m map[string]interface{}, keys []string, value interface{}) {
+		if len(keys) == 1 {
+			m[keys[0]] = value
+			return
+		}
+		key := keys[0]
+		if _, exists := m[key]; !exists {
+			m[key] = make(map[string]interface{})
+		} else {
+			if _, ok := m[key].(map[string]interface{}); !ok {
+				m[key] = make(map[string]interface{})
+			}
+		}
+		nestedMap := m[key].(map[string]interface{})
+		mergeNestedMap(nestedMap, keys[1:], value)
+	}
```

Comment:
I'm not sure about this function. First, it will overwrite any values with `map[string]any` if the value exists but it happens to not be a map[string]any. Second, it won't actually merge any maps. It will take the second one.

For example:

```go
	m1 := map[string]any{
		"m": map[string]any{
			"2.1": "value1",
		},
	}

	m2 := map[string]any{
		"m": map[string]any{
			"1": "value2",
		},
	}

	r := mergeMaps(m1, m2)

	fmt.Println(r)
```

Will result in `map[m:map[1:value2]]`. So the value1 is completely gone and `value2` is preserved.

Is that your intention? Because that looks like a bug if the intent is to merge two maps with potential sub maps defined by dots...?

Reply:
Even if I use just plain `2` as value, the value disappears.

```go
	m1 := map[string]any{
		"m": map[string]any{
			"2": "value1",
		},
	}

	m2 := map[string]any{
		"m": map[string]any{
			"1": "value2",
		},
	}
```

So there is definitely a bug in the `mergeMaps` not just this nested part.

Reply:
Ah yes. That's because down below the two loops. The second loop will just overwrite whatever key the first loop added. Is that intentional?

Reply:
Oh right, I see. So this will only work if the first key in the map contains `.`. It won't work on anything below that.

So like

```go
	m1 := map[string]any{
		"1.2.3": map[string]any{
			"3.2": "value1",
		},
	}
```

1.2.3 is mapified, but `3.2.` is not. I guess that's the intention?

Reply:
So regarding using this particular part. It will also overwrite this:

```go
	m1 := map[string]any{
		"nested": "merged",
	}

	m2 := map[string]any{
		"m2": map[string]any{
			"1": "value2",
		},
		"nested.4.4": map[string]any{
			"2": "value",
		},
	}

```

So nested in this section would be completely gone, unless it's also a map. In which case it _is_ preserved.